src/chainstate/stacks/index/cache.rs

```diff
@@ -0,0 +1,880 @@
+            TrieCache::Node256(ref mut state) => match node {
+                TrieNodeType::Node256(data) => {
+                    state.store_node(block_id, trieptr, TrieNodeType::Node256(data))
+                }
+                _ => {}
+            },
+        }
+    }
+
+    /// Store a node's hash to the cache.  `trieptr` must NOT be a backpointer
+    pub fn store_node_hash(&mut self, block_id: u32, trieptr: TriePtr, hash: TrieHash) {
+        assert!(!is_backptr(trieptr.id()));
+        match self {
+            TrieCache::Noop(_) => {
+                trace!(
+                    "Noop node hash cache store for ({},{:?},{})",
+                    block_id,
+                    &trieptr,
+                    &hash
+                );
+            }
+            TrieCache::Everything(ref mut state) => {
+                state.store_node_hash(block_id, trieptr, hash);
+            }
+            TrieCache::Node256(ref mut state) => match trieptr.id {
+                x if x == TrieNodeID::Node256 as u8 => {
+                    state.store_node_hash(block_id, trieptr, hash);
+                }
+                _ => {}
+            },
+        }
+    }
+
+    /// Load a block's hash, given its block ID.
+    pub fn load_block_hash(&mut self, block_id: u32) -> Option<T> {
+        match self {
+            TrieCache::Noop(ref mut state) => state.load_block_hash(block_id),
+            TrieCache::Everything(ref mut state) => state.load_block_hash(block_id),
+            TrieCache::Node256(ref mut state) => state.load_block_hash(block_id),
+        }
+    }
+
+    /// Store a block's ID and hash to teh cache.
+    pub fn store_block_hash(&mut self, block_id: u32, block_hash: T) {
+        match self {
+            TrieCache::Noop(ref mut state) => {
+                state.store_block_hash(block_id, block_hash);
+            }
+            TrieCache::Everything(ref mut state) => {
+                state.store_block_hash(block_id, block_hash);
+            }
+            TrieCache::Node256(ref mut state) => {
+                state.store_block_hash(block_id, block_hash);
+            }
+        }
+    }
+
+    /// Get an immutable reference to the block hash, given its ID
+    pub fn ref_block_hash(&self, block_id: u32) -> Option<&T> {
+        match self {
+            TrieCache::Noop(ref state) => state.ref_block_hash(block_id),
+            TrieCache::Everything(ref state) => state.ref_block_hash(block_id),
+            TrieCache::Node256(ref state) => state.ref_block_hash(block_id),
+        }
+    }
+
+    /// Get the block ID, given the block hash
+    pub fn load_block_id(&self, block_hash: &T) -> Option<u32> {
+        match self {
+            TrieCache::Noop(ref state) => state.load_block_id(block_hash),
+            TrieCache::Everything(ref state) => state.load_block_id(block_hash),
+            TrieCache::Node256(ref state) => state.load_block_id(block_hash),
+        }
+    }
+}
+
+#[cfg(test)]
```

Comment:
thought we were moving "test" code to special `test` files?

Reply:
That applies to `#[test]`-decorated functions.

Reply:
I agree with moving this test code -- this file is over 50% tests. Moving large test sets away from the code will make the code easier to navigate.